apps/passport-server/src/services/generic-issuance/SemaphoreGroupProvider.ts

```diff
@@ -0,0 +1,309 @@
+      const latestGroups =
+        await this.semaphoreHistoryDB.getLatestGroupsForPipeline(
+          this.pipelineId
+        );
+      const latestGroupMap = new Map(
+        latestGroups.map((group) => [group.groupId, group])
+      );
+
+      for (const groupConfig of this.groupConfigs) {
+        const historicalGroup = latestGroupMap.get(groupConfig.groupId);
+
+        this.groups.set(
+          groupConfig.groupId,
+          historicalGroup
+            ? deserializeSemaphoreGroup(
+                JSON.parse(historicalGroup.serializedGroup)
+              )
+            : new Group(
+                uuidToBigInt(groupConfig.groupId),
+                SEMAPHORE_GROUP_DEPTH
+              )
+        );
+      }
+
+      span?.setAttribute("groups_in_db_count", latestGroupMap.size);
+      span?.setAttribute("configured_groups_count", this.groupConfigs.length);
+
+      const configuredGroupIds = new Set(
+        this.groupConfigs.map((gc) => gc.groupId)
+      );
+      for (const groupId of latestGroupMap.keys()) {
+        if (!configuredGroupIds.has(groupId)) {
+          this.semaphoreHistoryDB.deleteGroup(this.pipelineId, groupId);
+        }
+      }
+    });
+  }
+
+  /**
+   * Updates the in-memory Semaphore groups using data on the current ticket
+   * set.
+   */
+  public async update(data: SemaphoreGroupTicketInfo[]): Promise<void> {
+    return traced(LOG_NAME, "update", async (span) => {
+      const groups = this.groupEmailsFromTicketData(data);
+
+      span?.setAttribute("pipeline_id", this.pipelineId);
+      span?.setAttribute("group_count", groups.size);
+
+      for (const [groupConfig, emails] of groups.entries()) {
+        await this.updateGroup(groupConfig, emails);
+      }
+    });
+  }
+
+  /**
+   * Updates a single group's in-memory representation, and save it to the DB
+   * if there have been any changes.
+   *
+   * We receive an array of { eventId, productId, email } objects, and
+   * from the emails we can look up Semaphore IDs via the consumer DB. We use
+   * the eventId/productId pairs to match configured Semaphore groups, and
+   * build up membership sets for each group. We then compare that to the
+   * in-memory representation, and then add or remove from the in-memory group
+   * representation as required.
+   */
+  private async updateGroup(
+    groupConfig: SemaphoreGroupConfig,
+    emails: string[]
+  ): Promise<void> {
+    return traced(LOG_NAME, "updateGroup", async (span) => {
+      const consumers =
+        emails.length > 0
+          ? await this.consumerDB.loadByEmails(this.pipelineId, emails)
+          : [];
+
+      const semaphoreIds = consumers.map((consumer) => consumer.commitment);
+      const group = this.groups.get(groupConfig.groupId);
+      if (!group) {
+        throw new Error(
+          `Could not find group ${groupConfig.groupId} on pipeline ${this.pipelineId}`
+        );
+      }
+
+      const { toAdd, toRemove } = this.calculateGroupChanges(
```

Comment:
`this.calculateGroupChanges` could potentially be extracted into a static function